Sources/ForageSDK/Foundation/Network/Collector.swift

```diff
@@ -323,6 +337,157 @@ class BasisTheoryWrapper: VaultCollector {
     }
 }
 
+// Wrapper class for Forage internal vault
+class ForageVaultWrapper: VaultCollector {
+    var customHeaders: [String: String] = [:]
+    let textElement: UITextField
+    
+    private let forageVaultConfig: ForageVaultConfig
+    private let logger: ForageLogger?
+    
+    init(textElement: UITextField, forageVaultConfig: ForageVaultConfig, logger: ForageLogger? = DatadogLogger(ForageLoggerConfig(prefix: "Rosetta"))) {
+        self.textElement = textElement
+        self.customHeaders = [:]
+        self.forageVaultConfig = forageVaultConfig
+        self.logger = logger
+    }
+    
+    func setCustomHeaders(headers: [String : String], xKey: [String : String]) {
+        customHeaders = headers
+    }
+    
+    func sendData<T: Decodable>(path: String, vaultAction: VaultAction, extraData: [String : Any], completion: @escaping (T?, ForageError?) -> Void) {
+        let semaphore = DispatchSemaphore(value: 0)
+        var pin = ""
+        
+        DispatchQueue.main.async {
+            guard let textElementValue = self.textElement.text else {
+                return completion(nil, CommonErrors.INCOMPLETE_PIN_ERROR)
+            }
+            pin = textElementValue
+            semaphore.signal()
+        }
+
+        // waiting for the value of the text field to be read from the main queue and saved in `pin`
+        semaphore.wait()
+
+        var body: [String: String] = ["pin": pin]
+
+        for (key, value) in extraData {
+            if key == tokenKey, let paymentMethodToken = value as? String {
+                do {
+                    let token = try getPaymentMethodToken(paymentMethodToken: paymentMethodToken)
+                    body[key] = token
+                } catch {
+                    logger?.critical(
+                        "Failed to send data to Rosetta proxy. Rosetta token not found on card",
+                        error: error,
+                        attributes: nil
+                    )
+                    return completion(nil, CommonErrors.UNKNOWN_SERVER_ERROR)
+                }
+            }
+        }
+        
+        let measurement = VaultProxyResponseMonitor.newMeasurement(vault: VaultType.forage, action: vaultAction)
+            .setPath(path)
+            .setMethod(.post)
+        
+        measurement.start()
```

Comment:
I've always moved the start command to the nearest possible line before the submit action (idk if this even matters after code compilation, but I still do it)

Reply:
[799b5d0](https://github.com/teamforage/forage-ios-sdk/pull/169/commits/799b5d07b8563b1ba67ccc8b9589dc7501c2b383)